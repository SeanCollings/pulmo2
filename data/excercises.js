import { DATA, STRENGTH_KEY, ENDURANCE_KEY, CUSTOM_KEY } from '.';

const STRENGTH = {
  id: STRENGTH_KEY,
  title: STRENGTH_KEY,
  icon: 'dumbbell',
  contents: DATA[STRENGTH_KEY],
};
const ENDURANCE = {
  id: ENDURANCE_KEY,
  title: ENDURANCE_KEY,
  icon: 'clock',
  contents: DATA[ENDURANCE_KEY],
};
const CUSTOM = {
  id: CUSTOM_KEY,
  title: CUSTOM_KEY,
  icon: 'wrench',
  contents: DATA[CUSTOM_KEY],
};

export const EXCERCISE_DATA = [STRENGTH, ENDURANCE, CUSTOM];
