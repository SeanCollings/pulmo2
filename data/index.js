export const STRENGTH_KEY = 'Strength';
export const ENDURANCE_KEY = 'Endurance';
export const CUSTOM_KEY = 'Custom';

export const DATA = {
  [STRENGTH_KEY]: [
    {
      id: 11,
      title: 'Strength Level I',
      cycles: 10,
      rest: 60,
      rounds: 4,
    },
    {
      id: 12,
      title: 'Strength Level II',
      cycles: 10,
      rest: 60,
      rounds: 5,
    },
    {
      id: 13,
      title: 'Strength Level III',
      cycles: 12,
      rest: 50,
      rounds: 5,
    },
  ],
  [ENDURANCE_KEY]: [
    {
      id: 21,
      title: 'Endurance Level I',
      cycles: 10,
      rest: 30,
      rounds: 4,
    },
    {
      id: 22,
      title: 'Endurance Level II',
      cycles: 12,
      rest: 20,
      rounds: 5,
    },
    {
      id: 23,
      title: 'Endurance Level III',
      cycles: 15,
      rest: 20,
      rounds: 5,
    },
  ],
  [CUSTOM_KEY]: [],
};

export const DEFAULT_EXCERCISE = DATA[STRENGTH_KEY][0];
