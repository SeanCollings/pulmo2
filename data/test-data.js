import { THEME } from '../context/settings-context';
import { DARK_MODE } from '../constants/constants';
import { SELECTED_LEVEL } from '../context/profile-context';
import { CUSTOM_TAG } from '../context/custom-excercise-context';
import { ACTIVITY_TAG, FAV_ACTIVITY_TAG } from '../context/history-context';
import { SELECTED_EXCERCISE } from '../context/excercise-context';

export const TEST_ALL_ACTIVITIES = 'test_all_activities';
export const TEST_KEY = 'test_key';
export const TEST_DATA = 'test_data';
export const TEST_TITLE = 'test_title';
export const TEST_MUTLIPLE = 'test_multiple';
export const TEST_MUTLIPLE_ID = 'test_multiple_id';
export const TEST_PRE_ID = 'test_pre_id';

//#region Custom Excercise ID Array
const test_customExcerciseIdArray = {
  [TEST_TITLE]: 'Custom Excercise IDs',
  [TEST_KEY]: CUSTOM_TAG,
  [TEST_DATA]: [1597490106660, 1597490120213],
};
//#endregion Custom Excercises ID Array

//#region Custom Excercises
const test_customExcercises = {
  [TEST_TITLE]: 'Custom Excercises',
  [TEST_MUTLIPLE_ID]: 'id',
  [TEST_PRE_ID]: CUSTOM_TAG,
  [TEST_DATA]: [
    {
      cycles: '12',
      id: 1597490106660,
      rest: '3',
      rounds: '4',
      title: 'Custom Excercise',
    },
    {
      cycles: '10',
      id: 1597490120213,
      rest: '1',
      rounds: '10',
      title: 'Many Rounds',
    },
  ],
};
//#endregion Custom Excercises

//#region Selected Excercise
const test_selected_excercise = {
  [TEST_TITLE]: 'Selected Excercise',
  [TEST_KEY]: SELECTED_EXCERCISE,
  [TEST_DATA]: [1597490106660, 'Custom'],
};
//#endregion Selected Excercise

//#region Activity ID Array
const test_activityIdArray = {
  [TEST_TITLE]: 'Activity IDs',
  [TEST_KEY]: ACTIVITY_TAG,
  [TEST_DATA]: [
    '2020-09-01T12:50:14.562Z',
    '2020-08-25T10:12:47.005Z',
    '2020-08-20T13:24:38.330Z',
    '2020-08-20T10:53:57.797Z',
    '2020-08-19T15:14:17.865Z',
    '2020-08-15T12:52:43.310Z',
    '2020-08-12T12:12:12.100Z',
    '2020-08-10T09:51:44.609Z',
    '2020-08-09T12:50:58.499Z',
    '2020-08-08T11:36:50.680Z',
    '2020-08-07T12:55:11.988Z',
    '2020-08-05T12:44:47.027Z',
    '2020-08-01T12:42:37.786Z',
    '2020-07-18T16:41:15.186Z',
    '2020-07-17T12:40:54.141Z',
    '2020-07-15T12:38:25.328Z',
    '2020-07-12T10:47:28.181Z',
    '2020-07-09T12:36:46.328Z',
    '2020-07-09T10:26:11.702Z',
    '2020-06-10T12:35:09.722Z',
    '2020-06-10T11:19:55.838Z',
  ],
};
//#endregion Activity ID Array

//#region Fav Activity ID Array
const test_favActivityIdArray = {
  [TEST_TITLE]: 'Favourite Activity IDs',
  [TEST_KEY]: FAV_ACTIVITY_TAG,
  [TEST_DATA]: [
    '2020-08-20T13:24:38.330Z',
    '2020-08-19T15:14:17.865Z',
    '2020-08-10T09:51:44.609Z',
    '2020-08-09T12:50:58.499Z',
    '2020-08-05T12:44:47.027Z',
    '2020-07-18T16:41:15.186Z',
    '2020-07-17T12:40:54.141Z',
    '2020-07-09T12:36:46.328Z',
    '2020-06-10T11:19:55.838Z',
  ],
};
//#endregion Fav Activity ID Array

//#region Profile
const test_profile_selected_level = {
  [TEST_TITLE]: 'Profile',
  [TEST_KEY]: SELECTED_LEVEL,
  [TEST_DATA]: 2,
};
//#endregion Profile

//#region Settings
const test_settings_theme = {
  [TEST_TITLE]: 'Settings',
  [TEST_KEY]: THEME,
  [TEST_DATA]: DARK_MODE,
};
//#endregion Settings

//#region All Activities
const test_all_activities = {
  [TEST_TITLE]: 'All Activities',
  [TEST_MUTLIPLE_ID]: 'date',
  [TEST_KEY]: TEST_ALL_ACTIVITIES,
  [TEST_DATA]: [
    {
      date: '2020-09-01T12:50:14.562Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise Here',
      },
      favourite: false,
      level: 2,
      results: [
        ['00:02', '00:03'],
        ['00:01', '00:03'],
        ['00:02', '00:03'],
        ['00:03', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-25T10:12:47.005Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 2,
      results: [
        ['00:03', '00:03'],
        ['00:03', '00:03'],
        ['00:03', '00:03'],
        ['00:03', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-20T13:24:38.330Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise 1',
      },
      favourite: true,
      level: 2,
      rating: 4,
      results: [
        ['00:07', '00:03'],
        ['00:02', '00:03'],
        ['00:07', '00:03'],
        ['00:04', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-20T10:53:57.797Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 2,
      results: [
        ['00:04', '00:03'],
        ['00:07', '00:03'],
        ['00:03', '00:03'],
        ['00:08', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-19T15:14:17.865Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      favourite: true,
      level: 2,
      results: [
        ['00:02', '00:03'],
        ['00:01', '00:03'],
        ['00:02', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-15T12:52:43.310Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      incomplete: ['placeholder'],
      level: 2.5,
      results: [['00:09', '00:01'], ['00:09', '00:01'], ['00:09', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-08-12T12:12:12.100Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      incomplete: ['placeholder'],
      level: 2.5,
      results: [['00:02', '00:01'], ['00:02', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-08-10T09:51:44.609Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      favourite: true,
      incomplete: ['feeling_lazy'],
      level: 2.5,
      results: [['00:02', '00:01'], ['00:02', '00:01'], ['00:02', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-08-09T12:50:58.499Z',
      excercise: {
        cycles: 12,
        id: 13,
        rest: 50,
        rounds: 5,
        title: 'Strength Level III',
      },
      favourite: true,
      level: 2.5,
      results: [
        ['00:05', '00:50'],
        ['00:05', '00:50'],
        ['00:06', '00:50'],
        ['00:06', '00:50'],
        ['00:04', ''],
      ],
      type: 'Strength',
    },
    {
      date: '2020-08-08T11:36:50.680Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 2.5,
      results: [
        ['00:02', '00:03'],
        ['00:02', '00:03'],
        ['00:02', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-07T12:55:11.988Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 3,
      results: [
        ['00:02', '00:03'],
        ['00:02', '00:03'],
        ['00:01', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-08-05T12:44:47.027Z',
      excercise: {
        cycles: 12,
        id: 22,
        rest: 20,
        rounds: 5,
        title: 'Endurance Level II',
      },
      favourite: true,
      level: 3,
      results: [
        ['00:03', '00:20'],
        ['00:04', '00:20'],
        ['00:08', '00:20'],
        ['00:05', '00:20'],
        ['00:05', ''],
      ],
      type: 'Endurance',
    },
    {
      date: '2020-08-01T12:42:37.786Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 3,
      results: [
        ['00:04', '00:03'],
        ['00:02', '00:03'],
        ['00:02', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-07-18T16:41:15.186Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Custom Many Rounds',
      },
      favourite: true,
      incomplete: ['out_of_breath'],
      level: 2.5,
      rating: 3,
      results: [['00:01', '00:01'], ['00:02', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-07-17T12:40:54.141Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      favourite: true,
      incomplete: ['other'],
      level: 3,
      results: [['00:08', '00:01'], ['00:05', '00:01'], ['00:05', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-07-15T12:38:25.328Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds 2',
      },
      level: 3,
      results: [
        ['00:02', '00:30'],
        ['00:02', '00:30'],
        ['00:03', '00:30'],
        ['00:03', '00:30'],
        ['00:04', '00:30'],
        ['00:04', '00:30'],
        ['00:03', '00:30'],
        ['00:02', '00:30'],
        ['00:03', '00:30'],
        ['00:03', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-07-12T10:47:28.181Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      incomplete: ['placeholder'],
      level: 3,
      results: [['00:03', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-07-09T12:36:46.328Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds Edited',
      },
      favourite: true,
      incomplete: ['placeholder'],
      level: 2.5,
      rating: 2,
      results: [
        ['00:03', '00:50'],
        ['00:04', '00:50'],
        ['00:05', '00:50'],
        ['00:02', '00:50'],
        [],
      ],
      type: 'Custom',
    },
    {
      date: '2020-07-09T10:26:11.702Z',
      excercise: {
        cycles: '10',
        id: 1597490120213,
        rest: '1',
        rounds: '10',
        title: 'Many Rounds',
      },
      incomplete: ['feeling_lazy'],
      level: 3.5,
      results: [['00:02', '00:01'], ['00:04', '00:01'], ['00:02', '00:01'], []],
      type: 'Custom',
    },
    {
      date: '2020-06-10T12:35:09.722Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      level: 3.5,
      results: [
        ['00:01', '00:03'],
        ['00:01', '00:03'],
        ['00:02', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
    {
      date: '2020-06-10T11:19:55.838Z',
      excercise: {
        cycles: '12',
        id: 1597490106660,
        rest: '3',
        rounds: '4',
        title: 'Custom Excercise',
      },
      favourite: true,
      level: 3.5,
      results: [
        ['00:02', '00:03'],
        ['00:02', '00:03'],
        ['00:03', '00:03'],
        ['00:02', ''],
      ],
      type: 'Custom',
    },
  ],
};
//#endregion All Activities

export default {
  test_customExcerciseIdArray,
  test_customExcercises,
  test_selected_excercise,
  test_activityIdArray,
  test_favActivityIdArray,
  test_profile_selected_level,
  test_settings_theme,
  test_all_activities,
};
