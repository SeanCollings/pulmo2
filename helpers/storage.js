import { AsyncStorage } from 'react-native';
import { APP_ID } from '../constants/constants';

export const getAsyncData = async (id) => {
  try {
    const value = await AsyncStorage.getItem(`${APP_ID}${id}`);
    if (value !== null) {
      return JSON.parse(value);
    } else {
      return null;
    }
  } catch (error) {
    throw new Error(`Failure retrieving ${id} with error: ${error}`);
  }
};

export const getMultiAsyncData = async (id, keys) => {
  try {
    const formattedKeys = keys.map((key) => `${APP_ID}${id}${key}`);
    const values = await AsyncStorage.multiGet(formattedKeys);

    if (values && !!values.length) return values;
    else {
      return null;
    }
  } catch (error) {
    throw new Error(`Failure retrieving from all keys with error: ${error}`);
  }
};

export const storeAsyncData = async (id, value) => {
  try {
    await AsyncStorage.setItem(`${APP_ID}${id}`, JSON.stringify(value));
    return true;
  } catch (error) {
    throw new Error('Failure saving: ' + error);
  }
};

export const mergeAsyncData = async (id, value) => {
  try {
    await AsyncStorage.mergeItem(`${APP_ID}${id}`, JSON.stringify(value));
    return true;
  } catch (error) {
    throw new Error('Failure merging: ' + error);
  }
};

export const removeAsyncData = async (id) => {
  try {
    await AsyncStorage.removeItem(`${APP_ID}${id}`);
    return true;
  } catch (error) {
    throw new Error('Failure removing: ' + error);
  }
};
