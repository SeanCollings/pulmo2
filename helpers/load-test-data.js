import testData, {
  TEST_KEY,
  TEST_TITLE,
  TEST_MUTLIPLE_ID,
  TEST_DATA,
  TEST_PRE_ID,
} from '../data/test-data';
import { getAsyncData, storeAsyncData } from './storage';

export const MESSAGE_UNKNOWN = 'unknown';
export const MESSAGE_PRE_SAVED = 'pre-saved';
export const MESSAGE_DONE = 'done';

export default async (title) => {
  try {
    const selectedData = Object.values(testData).find(
      (data) => data[TEST_TITLE] === title
    );

    if (!selectedData) return MESSAGE_UNKNOWN;

    if (!selectedData[TEST_MUTLIPLE_ID]) {
      const savedData = await getAsyncData(selectedData[TEST_KEY]);

      if (!savedData) {
        console.log('Saving data for:', selectedData[TEST_KEY]);
        await storeAsyncData(selectedData[TEST_KEY], selectedData[TEST_DATA]);
        return MESSAGE_DONE;
      } else {
        return MESSAGE_PRE_SAVED;
      }
    } else {
      const updatedSaves = await selectedData[TEST_DATA].reduce(
        async (acc, data) => {
          const newAcc = await acc;
          const id = selectedData[TEST_MUTLIPLE_ID];
          const preId = selectedData[TEST_PRE_ID] || '';
          const savedData = await getAsyncData(`${preId}${data[id]}`);

          if (!savedData) {
            console.log('Saving data for:', `${preId}${data[id]}`);
            newAcc.push(data[id]);
            await storeAsyncData(`${preId}${data[id]}`, data);
          }
          return newAcc;
        },
        []
      );

      if (!updatedSaves.length) return MESSAGE_PRE_SAVED;

      return `${MESSAGE_DONE} [${updatedSaves.length}]`;
    }
  } catch (err) {
    throw new Error(err);
  }
};
