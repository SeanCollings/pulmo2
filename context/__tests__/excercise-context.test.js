import React, { useContext } from 'react';
import { create } from 'react-test-renderer';
import { cleanup } from 'react-native-testing-library';

import ExcerciseContextProvider, {
  ExcerciseContext,
} from '../excercise-context';
import { getAsyncData } from '../../helpers/storage';
import { STRENGTH_KEY } from '../../data';

jest.mock('../../helpers/storage');

const MockComponent = () => {
  const context = useContext(ExcerciseContext);
  expect(context).toMatchSnapshot();
  return null;
};

describe('excercise-context', () => {
  afterEach(() => {
    cleanup();
    jest.clearAllMocks();
  });

  test('unit test', () => {
    getAsyncData.mockReturnValueOnce();

    let context = create(
      <ExcerciseContextProvider>
        <MockComponent />
      </ExcerciseContextProvider>
    );

    expect(context).toMatchSnapshot();
  });
});
