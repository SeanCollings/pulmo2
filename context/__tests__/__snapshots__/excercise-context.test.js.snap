// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`excercise-context unit test 1`] = `
Object {
  "selectedExcercise": Array [],
  "toggleExcercise": [Function],
}
`;

exports[`excercise-context unit test 2`] = `null`;
